gateway/ln-gateway/src/ng/register.rs

```diff
@@ -0,0 +1,149 @@
+use std::time::Duration;
+
+use fedimint_client::sm::{OperationId, State, StateTransition};
+use fedimint_client::DynGlobalClientContext;
+use fedimint_core::encoding::{Decodable, Encodable};
+use fedimint_core::task::sleep;
+use fedimint_ln_common::api::LnFederationApi;
+use fedimint_ln_common::LightningGateway;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use tracing::error;
+use url::Url;
+
+use super::GatewayClientContext;
+
+#[cfg_attr(doc, aquamarine::aquamarine)]
+/// State machine that registers the lightning gateway with
+/// the federation.
+///
+/// ```mermaid
+/// graph LR
+/// classDef virtual fill:#fff,stroke-dasharray: 5 5
+///
+///    Register -- register lightning gateway succeeded --> WaitForTTL
+///    Register -- register lightning gateway failed --> Register
```

Comment:
Would we ever need to exit this transition if we have continued `Register --> failed --> Register` loop?
The federation might have permanently gone offline (rugged), should we then transition to some recovery / archival states for this federation?

Reply:
Open to suggestions on how to handle this. Ideally we wouldn't keep trying over and over again like the PR does now, but we do want to try again in case the federation does come back online. Perhaps exponential backoff would be a good fit here.

Reply:
I added a `FailureBackoff` state so that sleeping on failure does not block other state machines. I also implemented exponential backoff (starting at 15 seconds and doubling) with a maximum of one day so that if the federation ever does come back, the gateway can be registered but it won't spam it with requests.